File: source_url_normalizer.py
"""Normalize source URLs before they are persisted or embedded."""

from __future__ import annotations

from urllib.error import HTTPError, URLError
from urllib.parse import parse_qsl, urlencode, urlparse, urlunparse
from urllib.request import Request, urlopen


SOUNDCLOUD_SHORT_HOSTS = {"on.soundcloud.com", "snd.sc"}
SOUNDCLOUD_HOSTS = {"soundcloud.com", "www.soundcloud.com", "m.soundcloud.com"}
SOUNDCLOUD_SHARE_PARAMS = {
    "c",
    "p",
    "ref",
    "si",
    "utm_campaign",
    "utm_medium",
    "utm_source",
}


def _host(source_url: str) -> str:
    try:
        parsed = urlparse(source_url)
        return parsed.hostname.lower() if parsed.hostname else ""
    except ValueError:
        return ""


def is_soundcloud_short_url(source_url: str) -> bool:
    """Return True for SoundCloud share links that redirect to a canonical URL."""

    return _host(source_url) in SOUNDCLOUD_SHORT_HOSTS


def is_soundcloud_url(source_url: str) -> bool:
    """Return True for canonical SoundCloud URLs."""

    return _host(source_url) in SOUNDCLOUD_HOSTS


def clean_soundcloud_url(source_url: str) -> str:
    """Strip SoundCloud sharing noise while preserving meaningful query params."""

    if not is_soundcloud_url(source_url):
        return source_url

    parsed = urlparse(source_url)
    hostname = parsed.hostname.lower() if parsed.hostname else parsed.netloc.lower()
    netloc = "soundcloud.com" if hostname in {"www.soundcloud.com", "m.soundcloud.com"} else hostname
    query = urlencode(
        [
            (key, value)
            for key, value in parse_qsl(parsed.query, keep_blank_values=True)
            if key.lower() not in SOUNDCLOUD_SHARE_PARAMS
        ],
        doseq=True,
    )

    return urlunparse(("https", netloc, parsed.path, "", query, ""))


def resolve_canonical_source_url(source_url: str, *, timeout: float = 8.0, opener=urlopen) -> str:
    """Resolve embeddable source URLs, currently focused on SoundCloud short links."""

    if is_soundcloud_url(source_url):
        return clean_soundcloud_url(source_url)

    if not is_soundcloud_short_url(source_url):
        return source_url

    resolved_url = source_url
    for method in ("HEAD", "GET"):
        request = Request(
            source_url,
            headers={"User-Agent": "Mozilla/5.0"},
            method=method,
        )

        try:
            with opener(request, timeout=timeout) as response:
                resolved_url = response.geturl()
            break
        except (HTTPError, URLError, TimeoutError, ValueError, OSError):
            continue

    if is_soundcloud_url(resolved_url):
        return clean_soundcloud_url(resolved_url)

    return source_url

File: apps/web/lib/jobs/source-url-normalizer.ts
const SOUNDCLOUD_SHORT_HOSTS = new Set(["on.soundcloud.com", "snd.sc"]);
const SOUNDCLOUD_HOSTS = new Set(["soundcloud.com", "www.soundcloud.com", "m.soundcloud.com"]);
const SOUNDCLOUD_SHARE_PARAMS = new Set([
  "c",
  "p",
  "ref",
  "si",
  "utm_campaign",
  "utm_medium",
  "utm_source",
]);

const getHost = (sourceUrl: string) => {
  try {
    return new URL(sourceUrl).hostname.toLowerCase();
  } catch {
    return "";
  }
};

export const isSoundCloudShortUrl = (sourceUrl: string) =>
  SOUNDCLOUD_SHORT_HOSTS.has(getHost(sourceUrl));

export const isSoundCloudUrl = (sourceUrl: string) =>
  SOUNDCLOUD_HOSTS.has(getHost(sourceUrl));

export const cleanSoundCloudUrl = (sourceUrl: string) => {
  if (!isSoundCloudUrl(sourceUrl)) {
    return sourceUrl;
  }

  const parsed = new URL(sourceUrl);
  const next = new URL(parsed);
  next.protocol = "https:";
  next.hostname = "soundcloud.com";
  next.hash = "";

  for (const key of [...next.searchParams.keys()]) {
    if (SOUNDCLOUD_SHARE_PARAMS.has(key.toLowerCase())) {
      next.searchParams.delete(key);
    }
  }

  return next.toString();
};

const fetchRedirectTarget = async (sourceUrl: string, method: "GET" | "HEAD", timeoutMs: number) => {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutMs);

  try {
    const response = await fetch(sourceUrl, {
      method,
      redirect: "follow",
      signal: controller.signal,
    });
    return response.url;
  } finally {
    clearTimeout(timeout);
  }
};

export const resolveCanonicalSourceUrl = async (
  sourceUrl: string,
  { timeoutMs = 8_000 }: { timeoutMs?: number } = {},
) => {
  if (isSoundCloudUrl(sourceUrl)) {
    return cleanSoundCloudUrl(sourceUrl);
  }

  if (!isSoundCloudShortUrl(sourceUrl)) {
    return sourceUrl;
  }

  for (const method of ["HEAD", "GET"] as const) {
    try {
      const resolvedUrl = await fetchRedirectTarget(sourceUrl, method, timeoutMs);
      if (isSoundCloudUrl(resolvedUrl)) {
        return cleanSoundCloudUrl(resolvedUrl);
      }
    } catch {
      // Fall through to the next method, then ultimately preserve the submitted URL.
    }
  }

  return sourceUrl;
};

export const resolveCanonicalSourceUrls = async (sourceUrls: string[]) => {
  const resolved = await Promise.all(sourceUrls.map((sourceUrl) => resolveCanonicalSourceUrl(sourceUrl)));
  return [...new Set(resolved)];
};
